### PR TITLE
Add initial support for base image handler

### PR DESCRIPTION
**Pull request by parth-07**

### CODE REVIEW

File: examples/plainPPM.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+int main(int argc, char* argv[])
+{
+    if (argc < 3)
+    {
+        std::cerr << "Usage: ./bin <in.png> <out.png>";
+        return 1;
+    }
+    std::unique_ptr<CIL::ImageHandler> plain_ppm_handler(new PlainPPMHandler());
+    CIL::AddImageHandler(std::move(plain_ppm_handler));
```
Comment (manorit2001) [inline examples/plainPPM.cpp:87]: do your template thingy here

File: include/CIL/PNG/PNGHandler.hpp [hunk 2]
```diff
@@ -1,42 +1,45 @@
...
-        class PNGHandler
+        class PNGHandler : public ImageHandler
         {
           public:
             /// Reads a PNG file and returns `CIL::ImageInfo` object for it.
             ///
             /// If the image uses palette color model then it is automatically
             /// converted to RGB color model.
-            static CIL::ImageInfo read(const char* filename);
+            CIL::ImageInfo readExact(const char* filename);
```
Comment (manorit2001) [inline include/CIL/PNG/PNGHandler.hpp:16]: either find a good name for it or remove it

File: lib/ImageHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#include <CIL/ImageHandler.hpp>
+
+#include <CIL/ImageInfo.hpp>
+
+#include <cassert>
```
Comment (manorit2001) [inline lib/ImageHandler.cpp:5]: extra spaces

File: lib/ImageInfo.cpp [hunk 4]
```diff
@@ -5,24 +5,53 @@
...
 #include <CIL/PNG/PNGHandler.hpp>
 #endif
+#include <CIL/ImageHandler.hpp>
 #include <CIL/PPM/PPMHandler.hpp>
+
+#include <memory>
+#include <vector>
 namespace CIL {
+    static std::vector<std::unique_ptr<ImageHandler>> registered_image_handlers;
+    // static std::vector<ImageHandler*> registered_image_handlers = {new PPMHandler(), new PNG::PNGHandler(), };
```
Comment (manorit2001) [inline lib/ImageInfo.cpp:15]: remove comment

**Review by manorit2001 — APPROVED** [review]
LGTM, The code is really nice, I have reviewed your code thoroughly and I couldn't understand a thing.

PS: I meant it, MERGE IT AT YOUR OWN RISK

**Comment by manorit2001** [comment]
STOP SPAMMING MY PR, it is **rude**

**Comment by parth-07** [comment]
> MERGE IT AT YOUR OWN RISK

THAT'S NOT A GOOD REVIEW AT ALL.

**Comment by parth-07** [comment]
> STOP SPAMMING MY PR, it is rude

Have to be me.

**Comment by manorit2001** [comment]
> > MERGE IT AT YOUR OWN RISK
> 
> THAT'S NOT A GOOD REVIEW AT ALL.

Works as long as you are to blame for merging it.

File: examples/plainPPM.cpp (hunk 1 above)
Comment (manorit2001) [reply examples/plainPPM.cpp:87]: Or, what do you think about - https://en.cppreference.com/w/cpp/memory/auto_ptr instead of unique_ptr?

Comment (parth-07) [reply examples/plainPPM.cpp:87]: `auto_ptr` is deprecated in C++ 11 and removed in C++17 sadly.

**Comment by manorit2001** [comment]
`Co-authored-by: Manorit Chawdhry <manorit2001@gmail.com>` and ready to merge

**Comment by manorit2001** [comment]
oh, add the files to automake too!

**Comment by manorit2001** [comment]
LGTM

**Pull request merged**